File: src/main/java/com/search/core/Query.java
package com.search.core;

import java.util.LinkedList;
import java.util.List;

public class Query {
    List<String> terms = new LinkedList<>();
    List<Operator> operators = new LinkedList<>();

    public void addTerm(String term) {
        this.terms.add(term);
    }

    public String getTerm() {
        if(this.terms.isEmpty()) {
            return null;
        }
        return this.terms.removeFirst();
    }

    public void addOperator(String strOp) {
        Operator operator = Operator.valueOf(strOp);
        this.operators.add(operator);
    }

    public boolean hasOperator() {
        return !this.operators.isEmpty();
    }

    public Operator getOperator() {
        if(this.operators.isEmpty()) {
            return null;
        }
        return this.operators.removeFirst();
    }
}

File: src/query.ts
enum EOperators {
    AND,
    OR,
    NOT,
};

export class Query {

    terms: string[];
    operator: EOperators[];

    constructor(term: string) {
        this.terms = [ term ];
        this.operator = [];
    }

    and(term: string) : this {
        this.terms.push(term);
        this.operator.push(EOperators.AND);
        return this;
    }

    or(term: string) : this {
        this.terms.push(term);
        this.operator.push(EOperators.OR);
        return this;
    }

    not(term: string) : this{
        this.terms.push(term);
        this.operator.push(EOperators.NOT);
        return this;
    }

}
